doc: add initial doc for consumers (#14)
pkgjs/support: doc/command-line-usage.md

# Using the packge support tool from the command line

There are 2 main uses cases for the package support tool

* Package Maintainers - document and communicate expectations to consumers
* Package Consumers - understand more about their key dependencies and where they 
  should invest.

The sections which follow document the usage/options of interest for each use
case.

## Package Consumers

> The support tool helpss package conumers review and understand the

Reviewer: ```suggestion
The support tool helps package consumers review and understand the
```

package support information provided by maintainers. See
[Package-support.md](https://github.com/nodejs/package-maintenance/blob/master/docs/PACKAGE-SUPPORT.md)

> which documents the suggested best practice and the spedific format of the

Reviewer: ```suggestion
which documents the suggested best practice and the specific format of the
```

package support information.

Package consumers use the `show` command to display the support information
provided by package maintainers.

It is understood that this is only the most basic support that can be provided
to the consumer but the current focus for the support tool is to provide features
to encourage/support maintainers to add the package support info to their packages.

> In the future more addvanced options can be added to the show command to provide 

Reviewer: ```suggestion
In the future more advanced options can be added to the show command to provide 
```

more deduplicated or focussed views of the information available.

> The general flow ia:s 

Reviewer: ```suggestion
The general flow is: 
```

* install node.js (version 10.x or later) and ensure it is on the path
* npm install the package support tool global (or optionally locally in which

>   case the `bin` directory need to be add the current path)

Reviewer 2: why not `npx` instead of globally installing?

Author: good point, I'll try that out and update

Author: @ljharb updated

* cd into the directory which contains the root package

> * execute `suport show`

Reviewer: ```suggestion
* execute `support show`
```

```shell
  @pkgjs/support(0.0.2) - { "versions": [ { "version": "*", "target": { "node": "supported" }, "response": { "type": "best-effort" }, "backing": { "hobby": "https://github.com/pkgjs/support" } } ] }
    @npmcli/arborist(0.0.0) - unknown
      @npmcli/installed-package-contents(1.0.5) - unknown
        npm-bundled(1.1.1) - unknown
        npm-normalize-package-bin(1.0.1) - unknown
      @npmcli/map-workspaces(0.0.0-pre.1) - unknown
        glob(7.1.6) - unknown
          fs.realpath(1.0.0) - unknown
          inflight(1.0.6) - unknown
            wrappy(1.0.2) - unknown

...

      y18n(4.0.0) - unknown
      yargs-parser(18.1.3) - unknown
        camelcase(5.3.1) - unknown
```

The command show the support information for the tree of packages starting from the support module itself. 
In this case it also ONLY shows support info for the `support` package as the support information has not
been added to packages yet.

The support info shown is:

```shell
  @pkgjs/support(0.0.2) - { "versions": [ { "version": "*", "target": { "node": "supported" }, "response": { "type": "best-effort" }, "backing": { "hobby": "https://github.com/pkgjs/support" } } ] }
```

> For the rest of the modules it simply shows `unknown` as the the packages do not yet provide it.

Reviewer: ```suggestion
For the rest of the modules it simply shows `unknown` as the packages do not yet provide it.
```

The package support tool prefers support information which is available locally with the package after installation.

> This is to allow offline use. The documented best practices for adding support info are written to maximmize

Reviewer: ```suggestion
This is to allow offline use. The documented best practices for adding support info are written to maximize
```

This is to allow offline use. The documented best practices for adding support info are written to maximize
the likelyhood that at reasonably up to date copy of the support information is provided as part of the npm
install.

If the consumer wishes to ignore the local information and always use the canonical version of the support information
they can add the `--canonical` option to the command. Instead of using the local support info, if available, it will

> display the URL from which the canonical information can be retrieved. This is the same behaviour when 

Reviewer 2: ```suggestion
display the URL from which the canonical information can be retrieved. This is the same behavior when 
```
unless we have a standard of using british spelling?

Author: No that's just the Canadian in me.  The standard is US english :)

add the `--fetch` option which will pull the remote support information when necesarry.

The following command/options will always show the most up to date package support information
at the cost of fetching from remote sites.
```
show --canonical --fetch
```

`--fetch` can also be used without `--canonical` in which case the remote support
information will only be pulled when it is not provided as part of the npm package.

Local support information may be available locally but outside of the package itself.
For example this may be the case if you are using mono-repo and sharing the support file.
In this case by default the support tool will not read/display files outside of the
packge itself to avoid potential security issues. In order to allow the tool to display
support info in these cases you can use the `--base-path` option to specify the top
most root from which packge support files can be read:

```shell
support  --base-path=${cwd}/../.. show
```

In the case where the required base path is not specified and the file is outside
of the package the tool will simply print the url form which the support information
can be obtained versus displaying it. For example:

```

> @pkgjs/support-show-local-escape-ok(0.0.1) - https://github.com/pkgjs/support/blob/master/my-support-info.jso

Reviewer: ```suggestion
@pkgjs/support-show-local-escape-ok(0.0.1) - https://github.com/pkgjs/support/blob/master/my-support-info.json
```

```

where the `support` entry was:

```json
"support": "../../fixtures/my-support-info.json",
```

## Package Maintainers

Package maintainers want to add and manage the support information for their modules.

Ultimately we'll want commands help generate the support info, but as a first step
we are provided the `validate` command which validates that the info is in the
right format/locations. 

To add support information a maintainer needs to:

> * add information to the package.json which specifices were to find the support info

Reviewer: ```suggestion
* add information to the package.json which specifies were to find the support info
```

* provide a file with the support info.

The `validate` command checks that both the information added to the package.json

> and the the file containing the support info (package-support.json by default) is valid.

Reviewer: ```suggestion
and the file containing the support info (package-support.json by default) is valid.
```

[Integration into package.json](https://github.com/nodejs/package-maintenance/blob/master/docs/PACKAGE-SUPPORT.md#integration-into-packagejson)
explains the options for providing support info and what goes into the package.json for
each case.

`validate` is run from the directory that contains the package.json for the package.
Depending the how the `support` section in the package.json is configured validate will:

* validate the format of the `support` section in the package.json conforms to the
documentation in [Integration into package.json](https://github.com/nodejs/package-maintenance/blob/master/docs/PACKAGE-SUPPORT.md#integration-into-packagejson)
* Calculate the file which cotains the support info
* If the file is available locally, use that data unless the `--canonical`
  option was specified.
* If the file is not available locally or `--canonical` was specified

>   * If --fetch was provided as an option, fetch the file and use thate data

Reviewer: ```suggestion
  * If --fetch was provided as an option, fetch the file and use that data
```

  otherwise simply print that the file was not resolved and the url from 
  where it can be retrieved.

If both the entry in the package.json and the contents of the file with the package support
information is validate you'll see the follwowing output:

```shell
Your support information is valid!
```

If the information support information is not avialable locally or you have specified
canonical without `--fetch` you'll see something like:

```shell
support info not resolved: https://github.com/pkgjs/support/blob/master/package-support.json
```

in this case the file is local to the package, but --canonical was specified. The same

> would be shown if the info in the package.json file pointed to a file in a different pacakge.

Reviewer: ```suggestion
would be shown if the info in the package.json file pointed to a file in a different package.
```

case.

Once you have node.js installed (version 10.x or later) you can
run the support command directly using npx or optionally npm install the
package and add the `bin` directory to your path. In either case
you can run the support tool using `npx support`.

## Package Consumers

The support tool helps package conumers review and understand the
package support information provided by maintainers. See
[Package-support.md](https://github.com/nodejs/package-maintenance/blob/master/docs/PACKAGE-SUPPORT.md)
which documents the suggested best practice and the specific format of the
package support information.

Package consumers use the `show` command to display the support information
provided by package maintainers.

It is understood that this is only the most basic support that can be provided
to the consumer. The current focus for the support tool is to provide features
to encourage/support maintainers to add the package support info to their packages.
In the future more advanced options can be added to the show command to provide 
more deduplicated or focussed views of the information available.

The general flow is:

> * install the package to display package support information

Author: ```suggestion
* install the package to display package support information for
```
